src/org/traccar/protocol/Arnavi4ProtocolDecoder.java

```diff
@@ -0,0 +1,187 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.protocol;
+
+import org.jboss.netty.buffer.ChannelBuffer;
+import org.jboss.netty.buffer.ChannelBuffers;
+import org.jboss.netty.channel.Channel;
+import org.traccar.BaseProtocolDecoder;
+import org.traccar.DeviceSession;
+import org.traccar.helper.Checksum;
+import org.traccar.model.Position;
+
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.Date;
+import java.util.LinkedList;
+import java.util.List;
+
+import static org.traccar.protocol.Arnavi4FrameDecoder.HEADER_START_SIGN;
+import static org.traccar.protocol.Arnavi4FrameDecoder.HEADER_VERSION_1;
+import static org.traccar.protocol.Arnavi4FrameDecoder.HEADER_VERSION_2;
+import static org.traccar.protocol.Arnavi4FrameDecoder.PACKAGE_START_SIGN;
+import static org.traccar.protocol.Arnavi4FrameDecoder.PACKAGE_END_SIGN;
+
+public class Arnavi4ProtocolDecoder extends BaseProtocolDecoder {
+
+    private static final byte RECORD_PING = 0x00;
+    private static final byte RECORD_DATA = 0x01;
+    private static final byte RECORD_TEXT = 0x03;
+    private static final byte RECORD_FILE = 0x04;
+    private static final byte RECORD_BINARY = 0x06;
+
+    private static final byte TAG_LATITUDE = 3;
+    private static final byte TAG_LONGITUDE = 4;
+    private static final byte TAG_COORD_PARAMS = 5;
+
+    public Arnavi4ProtocolDecoder(Arnavi4Protocol protocol) {
+        super(protocol);
+    }
+
+    private Position decodePosition(DeviceSession deviceSession, ChannelBuffer buf, int length, Date time) {
+
+        final Position position = new Position();
+        position.setProtocol(getProtocolName());
+        position.setDeviceId(deviceSession.getDeviceId());
+
+        position.setTime(time);
+
+        int readBytes = 0;
+        while (readBytes < length) {
+            short tag = buf.readUnsignedByte();
+            switch (tag) {
+                case TAG_LATITUDE:
+                    position.setLatitude(buf.readFloat());
+                    position.setValid(true);
+                    break;
+
+                case TAG_LONGITUDE:
+                    position.setLongitude(buf.readFloat());
+                    position.setValid(true);
+                    break;
+
+                case TAG_COORD_PARAMS:
+                    position.setCourse(buf.readUnsignedByte() * 2.0);
+                    position.setAltitude(buf.readUnsignedByte() * 10.0);
+                    byte satellites = buf.readByte();
+                    position.set(Position.KEY_SATELLITES, satellites & 0x0F + (satellites >> 4) & 0x0F); // gps+glonass
+                    position.setSpeed(buf.readUnsignedByte() * 1.852);
```

Comment:
Do you store the speed in knots or smth else? ARNAVI4 send the speed in knots by default.